File: List/169MajorityElement.py
class Solution:
    # Dictionary
    def majorityElement(self, nums: List[int]) -> int:
        n = len(nums)
        dictionary = defaultdict(int)
        
        for num in nums:
            dictionary[num] += 1;
            
        n = n//2
        
        for key, value in dictionary.items():
            if value > n:
                return key
            
        return 0
    
    '''
    # Moore Voting Algorithm
    def majorityElement(self, nums: List[int]) -> int:
        candidate = nums[0]
        count = 1
        
        for i in range(1, len(nums)):
            if count == 0:
                candidate = nums[i]
                
            count += 1 if candidate == nums[i] else -1
                
        return candidate
    '''

File: List/169MajorityElement.cpp
class Solution {
public:
    // Map 
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> map;
        
        for (int i = 0; i < n; i++)
            map[nums[i]]++;
        
        n = n/2;
        
        for (auto x : map)
        {
            if (x.second > n)
                return x.first;
        }
        
        return 0;
    } 
    
    /*
    // Moore Voting Algorithm
    int majorityElement(vector<int>& nums) {
        int candidate = nums[0];
        int count = 1;
        
        for (int i = 1; i < nums.size(); i++)
        {
            if (count  == 0) candidate = nums[i];       
            count += (candidate == nums[i]) ? 1 : -1;
        }
        
        return candidate;
    }
    */
};